numpy/core/src/umath/special_comparisons.cpp

```diff
@@ -0,0 +1,464 @@
+     */
+    for (int typenum = NPY_BYTE; typenum <= NPY_ULONGLONG; typenum++) {
+        spec->slots[0].pfunc = (void *)get_loop<comp>;
+
+        PyArray_DTypeMeta *Int = PyArray_DTypeFromTypeNum(typenum);
+
+        /* Register the spec/loop for both forward and backward direction */
+        spec->dtypes[0] = Int;
+        spec->dtypes[1] = PyInt;
+        int res = PyUFunc_AddLoopFromSpec((PyObject *)ufunc, spec);
+        if (res < 0) {
+            Py_DECREF(Int);
+            goto fail;
+        }
+        spec->dtypes[0] = PyInt;
+        spec->dtypes[1] = Int;
+        res = PyUFunc_AddLoopFromSpec((PyObject *)ufunc, spec);
+        Py_DECREF(Int);
+        if (res < 0) {
+            goto fail;
+        }
+    }
+
+    /*
+     * Install the promoter info to allow two Python integers to work.
+     */
+    return PyUFunc_AddLoop((PyUFuncObject *)ufunc, info, 0);
+
+    Py_DECREF(ufunc);
+    return 0;
+
+  fail:
+    Py_DECREF(ufunc);
+    return -1;
+}
+
+
+template<COMP...>
+struct add_loops;
+
+template<>
+struct add_loops<> {
+    int operator()(PyObject*, PyArrayMethod_Spec*, PyObject *) {
+        return 0;
+    }
+};
+
+
+template<COMP comp, COMP... comps>
+struct add_loops<comp, comps...> {
+    int operator()(PyObject* umath, PyArrayMethod_Spec* spec, PyObject *info) {
+        if (add_dtype_loops<comp>(umath, spec, info) < 0) {
+            return -1;
+        }
+        else {
+            return add_loops<comps...>()(umath, spec, info);
+        }
+    }
+};
+
+
+NPY_NO_EXPORT int
+init_special_int_comparisons(PyObject *umath)
+{
+    int res = -1;
+    PyObject *info = NULL, *promoter = NULL;
+    /* NOTE: This should receive global symbols? */
+    PyArray_DTypeMeta *Bool = PyArray_DTypeFromTypeNum(NPY_BOOL);
+
+    /* We start with the string loops: */
+    PyArray_DTypeMeta *dtypes[] = {NULL, NULL, Bool};
+    /*
+     * We only have one loop right now, the strided one.  The default type
+     * resolver ensures native byte order/canonical representation.
+     */
+    PyType_Slot slots[] = {
+        {_NPY_METH_get_loop, nullptr},
+        {NPY_METH_resolve_descriptors_raw, (void *)&resolve_descriptors_raw},
+        {0, NULL},
+    };
+
+    PyArrayMethod_Spec spec = {};
+    spec.name = "templated_pyint_comp";
+    spec.nin = 2;
+    spec.nout = 1;
+    spec.dtypes = dtypes;
+    spec.slots = slots;
+    spec.flags = NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    PyObject *dtype_tuple = PyTuple_Pack(3,
+            &PyArray_PyIntAbstractDType, &PyArray_PyIntAbstractDType, Bool);
+    if (dtype_tuple == NULL) {
+        goto finish;
+    }
+    promoter = PyCapsule_New(
+            (void *)&pyint_comparison_promoter, "numpy._ufunc_promoter", NULL);
+    if (promoter == NULL) {
+        Py_DECREF(promoter);
+        goto finish;
+    }
+    info = PyTuple_Pack(2, dtype_tuple, promoter);
+    Py_DECREF(dtype_tuple);
+    Py_DECREF(promoter);
+    if (info == NULL) {
+        goto finish;
+    }
+
+    /* All String loops */
```

Comment:
Again `string`???

Reply:
Sorry, I had started with a string ufuncs as they are also templated in C++ for all comparisons, and wasn't careful enough with removing all th leftovers.  Will look through more carefully.